include/groov/mmio_bus.hpp

```diff
@@ -5,32 +5,78 @@
 #include <async/just_result_of.hpp>
 #include <async/then.hpp>
 
+#include <groov/identity.hpp>
+
+#include <boost/mp11/algorithm.hpp>
+#include <boost/mp11/integral.hpp>
+#include <boost/mp11/list.hpp>
+
 #include <bit>
 #include <concepts>
 #include <cstdint>
 
 namespace groov {
 namespace detail {
-template <std::size_t NumBits> struct uint_of_width;
+using namespace boost::mp11;
+
+template <std::unsigned_integral T, std::size_t ByteOffset>
+struct subword_spec {
+    using subword_t = T;
+    constexpr static auto offset = ByteOffset;
 
-template <> struct uint_of_width<8> {
-    using type = std::uint8_t;
+    template <typename M>
+    constexpr static M mask =
+        static_cast<M>(std::numeric_limits<T>::max()) << (ByteOffset * 8);
 };
 
-template <> struct uint_of_width<16> {
-    using type = std::uint16_t;
+using subword_permuations =
+    mp_list<subword_spec<std::uint8_t, 0u>, subword_spec<std::uint8_t, 1u>,
+            subword_spec<std::uint8_t, 2u>, subword_spec<std::uint8_t, 3u>,
+            subword_spec<std::uint8_t, 4u>, subword_spec<std::uint8_t, 5u>,
+            subword_spec<std::uint8_t, 6u>, subword_spec<std::uint8_t, 7u>,
+
+            subword_spec<std::uint16_t, 0u>, subword_spec<std::uint16_t, 1u>,
+            subword_spec<std::uint16_t, 2u>, subword_spec<std::uint16_t, 3u>,
+            subword_spec<std::uint16_t, 4u>, subword_spec<std::uint16_t, 5u>,
+            subword_spec<std::uint16_t, 6u>,
+
+            subword_spec<std::uint32_t, 0u>, subword_spec<std::uint32_t, 1u>,
+            subword_spec<std::uint32_t, 2u>, subword_spec<std::uint32_t, 3u>,
+            subword_spec<std::uint32_t, 4u>,
+
+            subword_spec<std::uint64_t, 0u>>;
+
+template <std::unsigned_integral BaseType> struct fits_within {
+    constexpr static auto base_size = sizeof(BaseType);
+
+    template <typename S>
+    using fn = std::bool_constant<(sizeof(typename S::subword_t) + S::offset) <=
+                                  base_size>;
 };
 
-template <> struct uint_of_width<32> {
-    using type = std::uint32_t;
+template <typename HardwareInterface> struct aligned_for {
+    using iface = HardwareInterface;
+
+    template <typename S>
+    using fn = std::bool_constant<
+        (S::offset % iface::template alignment<typename S::subword_t>) == 0>;
 };
 
-template <> struct uint_of_width<64> {
-    using type = std::uint64_t;
+template <auto Mask, decltype(Mask) IdMask> struct subword_satisfies {
+    using base_type = decltype(Mask);
+
+    template <typename S> constexpr static auto calculate() -> bool {
+        constexpr base_type write_mask = S::template mask<base_type>;
+        constexpr base_type id_write_mask = write_mask & ~Mask;
+        constexpr bool subword_covers_mask = (Mask & write_mask) == Mask;
+        constexpr bool single_store_en =
+            (id_write_mask & IdMask) == id_write_mask;
+        return subword_covers_mask && single_store_en;
```

Comment:
**thought**: The names here are a bit confusing. The template parameter `Mask` is elsewhere called the write mask, right? If I understand what is happening here, this might be clearer?

```cpp
constexpr base_type subword_mask = S::template mask<base_type>;
constexpr bool subword_covers_mask =  (Mask & subword_mask) == Mask;
constexpr bool subword_covered = ((Mask | IdMask) & subword_mask) == subword_mask;
return subword_covers_mask and subword_covered;
```